packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,293 @@
+// Below here is copied from uplot (MIT License)
+// https://github.com/leeoniya/uPlot/blob/master/src/utils.js#L325
+// This avoids needing to import uplot into the data package
+//--------------------------------------------------------------------------------
+
+// Copied from uplot
+type AlignedData = [number[], ...Array<Array<number | null>>];
+
+// nullModes
+const NULL_REMOVE = 0; // nulls are converted to undefined (e.g. for spanGaps: true)
+const NULL_RETAIN = 1; // nulls are retained, with alignment artifacts set to undefined (default)
+const NULL_EXPAND = 2; // nulls are expanded to include any adjacent alignment artifacts
+
+type JoinNullMode = number; // NULL_IGNORE | NULL_RETAIN | NULL_EXPAND;
+
+// sets undefined values to nulls when adjacent to existing nulls (minesweeper)
+function nullExpand(yVals: Array<number | null>, nullIdxs: number[], alignedLen: number) {
+  for (let i = 0, xi, lastNullIdx = -1; i < nullIdxs.length; i++) {
+    let nullIdx = nullIdxs[i];
+
+    if (nullIdx > lastNullIdx) {
+      xi = nullIdx - 1;
+      while (xi >= 0 && yVals[xi] == null) {
+        yVals[xi--] = null;
+      }
+
+      xi = nullIdx + 1;
+      while (xi < alignedLen && yVals[xi] == null) {
+        yVals[(lastNullIdx = xi++)] = null;
+      }
+    }
+  }
+}
+
+// nullModes is a tables-matched array indicating how to treat nulls in each series
+function join(tables: AlignedData[], nullModes: number[][]) {
+  if (tables.length === 1) {
+    return tables[0];
+  }
+
+  let xVals = new Set<number>();
+
+  for (let ti = 0; ti < tables.length; ti++) {
+    let t = tables[ti];
+    let xs = t[0];
+    let len = xs.length;
+
+    for (let i = 0; i < len; i++) {
+      xVals.add(xs[i]);
+    }
+  }
+
+  let data = [Array.from(xVals).sort((a, b) => a - b)];
+
+  let alignedLen = data[0].length;
+
+  let xIdxs = new Map();
+
+  for (let i = 0; i < alignedLen; i++) {
+    xIdxs.set(data[0][i], i);
+  }
+
+  for (let ti = 0; ti < tables.length; ti++) {
+    let t = tables[ti];
+    let xs = t[0];
+
+    for (let si = 1; si < t.length; si++) {
+      let ys = t[si];
+
+      let yVals = Array(alignedLen).fill(undefined);
+
+      let nullMode = nullModes ? nullModes[ti][si] : NULL_RETAIN;
+
+      let nullIdxs = [];
+
+      for (let i = 0; i < ys.length; i++) {
+        let yVal = ys[i];
+        let alignedIdx = xIdxs.get(xs[i]);
+
+        if (yVal == null) {
+          if (nullMode !== NULL_REMOVE) {
+            yVals[alignedIdx] = yVal;
+
+            if (nullMode === NULL_EXPAND) {
+              nullIdxs.push(alignedIdx);
+            }
+          }
+        } else {
+          yVals[alignedIdx] = yVal;
+        }
+      }
+
+      nullExpand(yVals, nullIdxs, alignedLen);
+
+      data.push(yVals);
+    }
+  }
+
+  return data;
+}
+
+// Quick test if the first and last points look to be ascending
+export function isLikelyAscendingVector(data: Vector): boolean {
```

Comment:
this util function is not join-specific, probably pull out of this file?

Reply:
it is still a bit too specific to be generally exported -- if we have a use elsewhere we can move it